### ISSUE TITLE
Embedded SBoM Formats section

### ISSUE THREAD

**Issue by robert-ami** [hughsie/uefi-sbom-best-practices#5]
Our question here is why coSWID and not CycloneDX or SPDX?

 As I understand, the key advantage is compactness; however, if requirement of SBOM per PE image is lifted, size constraint is also alleviated. Having that, usage of CycloneDX or SPDX  instead of CoSWID if preferable as more mainstream, directly human consumable, and mandated by the executive order 14028 formats. The last point is take from the document(see below): “To comply with Executive Order 14028, OEM vendors SHOULD also publish either the SPDX or CycloneDX SBoM”. 

In other words, CycloneDX or SPDX has to be supported anyway, so why also spend time on supporting CoSWID?

**Comment by hughsie** [comment]
> Our question here is why coSWID and not CycloneDX or SPDX?

Okay; format comparison first.

SPDX is a non-starter as I've received significant pushback that $250 to read the specification is an insane barrier to entry for anyone that's not a company generating thousands of dollars in revenue. $250 to read is okay for AMI, but not okay for a random open hardware project -- or an ISV with super-small margins.

CycloneDX doesn't actually give us enough information for firmware; it's "lossy" in the context of firmware. You can't go from SWID to CycloneDX to SPDX -- there's a impedance mismatch in the spec and you have to fudge a square peg into a round hole. Nestable CycloneDX to flat SPDX, `software-creator` to missing `maintainer` for example. Check out `uswid/format_*.py` and you can see how SWID is used as a superset format that can export to either CycloneDX or SPDX.

Also note that coSWID is DTMF so no $250 fee to implement.

Implementation next. You don't probably don't want a JSON or YAML parser in your firmware image, and you've probably already got CBOR for other certificate things. So if you want to show the *current* SBoM as part of the BIOS splash or in the BIOS config itself you need to parse the document, possibly from a known offset. CBOR can be progressively streamed and filtered, something in an export format like XML has to be sanitized and then loaded _as_ a DOM or by _building_ a DOM. We also might need to merge the SBoM at runtime, for instance it wouldn't make sense to ship a pre-generated HTML in the SPI chip just like it doesn't make sense to ship SWID _XML_ -- it's an export format not an internal representation.

We've also got firmware shipping right now that uses coSWID and I'm not super happy with starting back at the beginning seeing as we're so far behind. If UEFI was leading this SBoM initiative 3 years ago I think the "_we want to use something else_" sentiment would be much easier than now, but the reality on the ground is that Coreboot has been shipping this stuff for over a year and peripheral vendors have been adding this a ISV requirement for consumer products releasing in 2024. 3 years ago when we started uSWID, CycloneDX wasn't at v1.4 (which is the bare minimum for what we need in firmware) and SPDX wasn't an ISO format, only in 2022 it gained enough metadata compatibility (v2.3) to be used for anything other than _licence compliance_.

> usage of CycloneDX or SPDX instead of CoSWID if preferable as more mainstream, directly human consumable

Reading a BIOS IFD partition from your SPI device is not at all human consumable, and I don't think EO 14028 is talking about the _encoding_ of the embedded data anywhere. What I take the EO and Cyber Trust Mark to mean is that the OEM has to provide the computer-readable asset as either CycloneDX or SPDX for the hardware model **on the OEM webpage**. To be human readable, implies an HTML page (as I can't parse JSON/XML in my head) -- both of what the LVFS provides. We store the (perhaps multiple) coSWID compressed blocks in the firmware image, extract them and make them available as downloadable CycloneDX, SPDX and SWID XML documents and also provide an HTML page (with clickable links) for human use. EO 14028 doesn't stop me **storing** the SBoM in the firmware as a GIF image (which would be ridiculous too), it just mandates what is made available, i.e. **published** to the end user.

> CycloneDX or SPDX has to be supported anyway, so why also spend time on supporting CoSWID

I think what I wrote above justifies that question. I'm not against uSWID-prefixed SPDX XML and YAML and uSWID-prefixed CycloneDX on a technical level, but in the context of implementation saying "_if you want to read the SBoM out of this image you need to support 3 different encoding formats using 3 different parsers_" is just crazy. Having three different import formats and three different export formats gives you a complexity of *9* (and more, if you want to support different versions of SPDX and CycloneDX) and given that it's a lossy conversion from one format to another makes actually using this under-defined data encapsulation in real projects almost impossible.

**Comment by hughsie** [comment]
> I'm reading it without paying anything

That might be the same thing (or something else entirely?), but the official standard is found here: https://www.iso.org/standard/81870.html -- take a look at the press release here: https://www.linuxfoundation.org/press/featured/spdx-becomes-internationally-recognized-standard-for-software-bill-of-materials
